engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageState.java

```diff
@@ -336,17 +337,28 @@ public StoredMessage getMessage(final long messageKey) {
   }
 
   @Override
-  public void visitMessagesWithDeadlineBefore(
-      final long timestamp, final ExpiredMessageVisitor visitor) {
-    deadlineColumnFamily.whileTrue(
-        ((compositeKey, zbNil) -> {
-          final long deadline = compositeKey.first().getValue();
-          if (deadline <= timestamp) {
-            final long messageKey = compositeKey.second().inner().getValue();
-            return visitor.visit(messageKey);
-          }
-          return false;
-        }));
+  public boolean visitMessagesWithDeadlineBeforeTimestamp(
+      final long timestamp, final Index startAt, final ExpiredMessageVisitor visitor) {
+    final var stoppedByVisitor = new MutableBoolean(false);
+    if (startAt == null) {
```

Comment:
Seems like a reasonable change to me :+1: I'd suggest to do this here directly though, no need to go through a separate PR just because zb-db is ZDP responsibility.

Reply:
👍 Makes sense to me too, but I don't see it as a hard requirement. @oleschoenburg Do you want to pick that up?

Reply:
![image](https://user-images.githubusercontent.com/1379753/222666227-7ed9d208-2a14-4618-a32f-39c6a1213838.png)

Sure, I'll do it :+1:

Reply:
👍 Thanks @oleschoenburg 

✅ With #11937 merged, and this pull request rebased on top of `stable/8.1`, I have added 79193d3 to resolve this.